testing/file_test/test_file.cpp

```diff
@@ -133,30 +180,49 @@ static auto ReplaceLspKeywordAt(std::string* content, size_t keyword_pos,
   llvm::StringRef body = body_start.take_front(body_end);
   auto [method_or_id, extra_content] = body.split(":");
 
-  // Form the JSON.
-  std::string json = llvm::formatv(R"({{"jsonrpc": "2.0", "{0}": "{1}")",
-                                   method_or_id_label, method_or_id);
-  if (use_call_id) {
-    // Omit quotes on the ID because we know it's an integer.
-    json += llvm::formatv(R"(, "id": {0})", ++lsp_call_id);
-  }
+  llvm::json::Value parsed_extra_content = nullptr;
   if (!extra_content.empty()) {
-    json += ",";
-    if (extra_content_label.empty()) {
-      if (!extra_content.starts_with("\n")) {
-        json += " ";
-      }
-      json += extra_content;
-    } else {
-      json += llvm::formatv(R"( "{0}": {{{1}})", extra_content_label,
-                            extra_content);
+    std::string bracketed_extra_content =
+        llvm::formatv("{{{0}}", extra_content);
+    auto parse_result = llvm::json::parse(bracketed_extra_content);
+    if (auto err = parse_result.takeError()) {
+      return ErrorBuilder() << "Error parsing extra content: " << err;
+    }
+    parsed_extra_content = std::move(*parse_result);
+    CARBON_CHECK(parsed_extra_content.kind() == llvm::json::Value::Object);
+    if (extra_content_label == "params" &&
+        method_or_id == "textDocument/didOpen") {
+      CARBON_RETURN_IF_ERROR(
+          AutoFillDidOpenParams(parsed_extra_content.getAsObject(), splits));
     }
   }
-  json += "}";
+
+  // Form the JSON.
+  std::string buffer;
+  llvm::raw_string_ostream buffer_os(buffer);
+  llvm::json::OStream json(buffer_os);
+
+  json.object([&] {
+    json.attribute("jsonrpc", "2.0");
+    json.attribute(method_or_id_label, method_or_id);
+
+    if (use_call_id) {
+      json.attribute("id", ++lsp_call_id);
+    }
+    if (parsed_extra_content != nullptr) {
+      if (!extra_content_label.empty()) {
+        json.attribute(extra_content_label, parsed_extra_content);
+      } else {
+        for (const auto& [key, value] : *parsed_extra_content.getAsObject()) {
+          json.attribute(key, value);
+        }
+      }
+    }
+  });
 
   // Add the Content-Length header. The `2` accounts for extra newlines.
   auto json_with_header =
-      llvm::formatv("Content-Length: {0}\n\n{1}\n", json.size() + 2, json)
+      llvm::formatv("Content-Length: {0}\n\n{1}\n", buffer.size() + 2, buffer)
```

Comment:
```suggestion
      llvm::formatv("Content-Length: {0}\n\n{1}\n", buffer.size() + 2, buffer.TakeStr())
```

Continuing on RawStringOstream comment -- note this works because of the order of params

Reply:
Done, made a small change to calculate buffer.size() outside the call expression, as IIUC the order of evaluation for buffer.size() and buffer.TakeStr() is not guaranteed.